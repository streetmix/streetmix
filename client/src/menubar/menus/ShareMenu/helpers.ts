import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
