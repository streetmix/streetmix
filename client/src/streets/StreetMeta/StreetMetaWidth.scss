.street-width {
  display: inline-flex;
  align-items: center;
}

.street-width .street-meta-icon svg {
  /* Radix UI optimized width/height */
  width: 15px;
  height: 15px;
}

.street-width .menu-carat-down {
  transform: scale(0.85);
}

.street-width-over {
  margin-left: 0.25em;
  color: var(--warning-color);

  [dir="rtl"] & {
    margin-right: 0.25em;
    margin-left: 0;
  }
}

.street-width-under {
  margin-left: 0.25em;
  color: rgb(96 96 96);
  transition: color var(--skybox-transition);

  [dir="rtl"] & {
    margin-right: 0.25em;
    margin-left: 0;
  }
}

// Invert the UI text color when there is a dark background
body.dark-skybox-invert-ui {
  .street-width-under {
    color: #a0a0a0;
  }
}

[data-radix-popper-content-wrapper] {
  z-index: 900 !important;
}

.dropdown-menu-content {
  --menu-border-radius: var(--border-radius);
  --menu-box-shadow: var(--medium-box-shadow);

  padding: 5px;
  background-color: white;
  transform-origin: 50% 0;
  border-radius: var(--menu-border-radius);
  box-shadow: var(--menu-box-shadow);
  line-height: 1.4;
  animation-duration: 100ms;
  will-change: transform, opacity;
}

.dropdown-menu-content[data-side="bottom"] {
  animation-name: swing-down;
}

.dropdown-menu-item,
.dropdown-menu-checkbox-item,
.dropdown-menu-radio-item {
  user-select: none;
  outline: none;
  padding: 0.5em 1em 0.5em 1.5em;
  text-decoration: none;
  color: black;
  border-radius: var(--border-radius);
  cursor: pointer;
}

.dropdown-menu-item[data-disabled],
.dropdown-menu-checkbox-item[data-disabled],
.dropdown-menu-radio-item[data-disabled] {
  color: #999; // TODO: gray color
  pointer-events: none;
}

.dropdown-menu-item[data-highlighted],
.dropdown-menu-checkbox-item[data-highlighted],
.dropdown-menu-radio-item[data-highlighted] {
  background-color: var(--interactive-element-hover-color);
}

.dropdown-menu-checkbox-item[data-state="checked"],
.dropdown-menu-radio-item[data-state="checked"] {
  font-weight: 500;
}

.dropdown-menu-label {
  padding: 0.5em 1em 0.5em 1.5em;
  font-weight: 500;
  color: var(--color-turquoise-700);
}

.dropdown-menu-item-indicator {
  position: absolute;
  left: 0.65em;
  margin-top: 3px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  color: var(--color-turquoise-500);
}

.dropdown-menu-arrow {
  fill: white;
}

@keyframes swing-down {
  from {
    opacity: 0;
    transform: rotateX(-90deg);
    transition:
      transform 100ms,
      opacity 100ms;
  }

  to {
    opacity: 1;
    transform: rotateX(0);
  }
}
