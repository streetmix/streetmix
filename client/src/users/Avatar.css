.avatar {
  display: inline-block;
  position: relative;
  background-size: 100%;
  background-color: rgba(127 127 127 / 10%);
  width: 20px;
  height: 20px;
  border-radius: 50%;
  vertical-align: middle;
  margin-left: 3px;
  margin-right: 5px;
  overflow: hidden;

  /* Avatar images are loaded in an object element to allow for fallbacks
     when image source is broken */
  object {
    width: 100%;
    height: 100%;
  }

  /* Should only display as fallback when object source is broken */
  img {
    border-radius: 50%;
    width: 100%;
    height: 100%;
    border: 1px solid white;
  }
}
