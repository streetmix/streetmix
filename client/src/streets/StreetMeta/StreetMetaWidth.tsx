import React from 'react'
import { FormattedMessage, useIntl } from 'react-intl'
import * as DropdownMenu from '@radix-ui/react-dropdown-menu'

import { useSelector, useDispatch } from '~/src/store/hooks'
import { updateStreetWidthAction as updateStreetWidth } from '~/src/store/actions/street'
import Icon from '~/src/ui/Icon'
import {
  prettifyWidth,
  processWidthInput,
  convertImperialMeasurementToMetric
} from '~/src/util/width_units'
import {
  SETTINGS_UNITS_IMPERIAL,
  SETTINGS_UNITS_METRIC
} from '~/src/users/constants'
import { updateUnits } from '~/src/users/localization'
import {
  MIN_CUSTOM_STREET_WIDTH,
  MAX_CUSTOM_STREET_WIDTH,
  MIN_CUSTOM_STREET_WIDTH_IMPERIAL,
  MAX_CUSTOM_STREET_WIDTH_IMPERIAL
} from '../constants'
import { normalizeStreetWidth } from '../width'
import StreetMetaItem from './StreetMetaItem'

import type { StreetJsonExtra } from '@streetmix/types'
import './StreetMetaWidth.scss'

const DEFAULT_STREET_WIDTHS_IMPERIAL = [40, 60, 80].map(
  convertImperialMeasurementToMetric
)
const DEFAULT_STREET_WIDTHS_METRIC = [12, 18, 24]

function StreetMetaWidthNew (): React.ReactElement | null {
  const street = useSelector((state) => state.street)
  const editable = useSelector(
    (state) => !state.app.readOnly && state.flags.EDIT_STREET_WIDTH.value
  )
  const locale = useSelector((state) => state.locale.locale)
  const dispatch = useDispatch()
  const width = prettifyWidth(street.width, street.units, locale)

  // A title attribute is provided only when street width is editable
  const intl = useIntl()
  const tooltip = intl.formatMessage({
    id: 'tooltip.street-width',
    defaultMessage: 'Change width of the street'
  })

  // Apply a class when street width is editable for styling
  let className = 'street-width'
  if (editable) {
    className += ' street-width-editable'
  }

  const { units, occupiedWidth } = street
  const defaultWidths =
    units === SETTINGS_UNITS_IMPERIAL
      ? DEFAULT_STREET_WIDTHS_IMPERIAL
      : DEFAULT_STREET_WIDTHS_METRIC

  function renderStreetWidthRemaining (
    { remainingWidth, units }: StreetJsonExtra,
    locale: string
  ): React.ReactElement | null {
    const width = prettifyWidth(Math.abs(remainingWidth), units, locale)

    if (remainingWidth > 0) {
      return (
        <span className="street-width-under">
          <FormattedMessage
            id="width.under"
            defaultMessage="({width} room)"
            values={{ width }}
          />
        </span>
      )
    } else if (remainingWidth < 0) {
      return (
        <span className="street-width-over">
          <FormattedMessage
            id="width.over"
            defaultMessage="({width} over)"
            values={{ width }}
          />
        </span>
      )
    }

    return null
  }

  function handleWidthChange (value: string): void {
    const newWidth = Number.parseFloat(value)

    // Do nothing if the selection is the original width
    if (newWidth === street.width) return

    // Change width to the desired selection
    if (newWidth) {
      void dispatch(updateStreetWidth(newWidth))
    }
  }

  function handleUnitChange (value: string): void {
    const selection = Number.parseFloat(value)
    updateUnits(selection)
  }

  function handleWidthEntry (): void {
    const minValue =
      units === SETTINGS_UNITS_IMPERIAL
        ? MIN_CUSTOM_STREET_WIDTH_IMPERIAL
        : MIN_CUSTOM_STREET_WIDTH
    const maxValue =
      units === SETTINGS_UNITS_IMPERIAL
        ? MAX_CUSTOM_STREET_WIDTH_IMPERIAL
        : MAX_CUSTOM_STREET_WIDTH
    const promptValue = normalizeStreetWidth(occupiedWidth, units)
    const promptString = intl.formatMessage(
      {
        id: 'prompt.new-width',
        defaultMessage: 'New street width (from {minWidth} to {maxWidth}):'
      },
      {
        minWidth: prettifyWidth(minValue, units, locale),
        maxWidth: prettifyWidth(maxValue, units, locale)
      }
    )
    const inputWidth = window.prompt(
      promptString,
      prettifyWidth(promptValue, units, locale)
    )

    // Ignore if input is cancelled or blank
    if (inputWidth === null || inputWidth === '') {
      return
    }

    const newWidth = normalizeStreetWidth(
      processWidthInput(inputWidth, units),
      units
    )

    void dispatch(updateStreetWidth(newWidth))
  }

  const DefaultWidthOptions = (): React.ReactElement[] =>
    defaultWidths.map((width) => (
      <DropdownMenu.RadioItem
        className="dropdown-menu-radio-item"
        value={width.toString()}
        key={width}
      >
        <DropdownMenu.ItemIndicator className="dropdown-menu-item-indicator">
          <Icon name="check" />
        </DropdownMenu.ItemIndicator>
        {prettifyWidth(width, units, locale)}
      </DropdownMenu.RadioItem>
    ))

  return (
    <DropdownMenu.Root>
      <DropdownMenu.Trigger asChild={true}>
        <StreetMetaItem
          className={className}
          isEditable={editable}
          icon={<Icon name="ruler" />}
          tooltip={tooltip}
        >
          <span className="underline">
            <FormattedMessage
              id="width.label"
              defaultMessage="{width} width"
              values={{ width }}
            />
          </span>
          {renderStreetWidthRemaining(street, locale)}
          {editable && <Icon name="chevron-down" className="menu-carat-down" />}
        </StreetMetaItem>
      </DropdownMenu.Trigger>

      <DropdownMenu.Portal>
        <DropdownMenu.Content className="dropdown-menu-content">
          <DropdownMenu.Label className="dropdown-menu-label">
            <FormattedMessage
              id="width.occupied"
              defaultMessage="Occupied width"
            />
          </DropdownMenu.Label>
          <DropdownMenu.Item className="dropdown-menu-item" disabled={true}>
            {prettifyWidth(occupiedWidth, units, locale)}
          </DropdownMenu.Item>

          <DropdownMenu.Separator className="menu-separator" />

          <DropdownMenu.Label className="dropdown-menu-label">
            <FormattedMessage
              id="width.building"
              defaultMessage="Building-to-building width"
            />
          </DropdownMenu.Label>

          <DropdownMenu.RadioGroup
            value={street.width.toString()}
            onValueChange={handleWidthChange}
          >
            <DefaultWidthOptions />

            {/* If the street width doesn't match any of the default widths,
            // render another choice representing the current width */}
            {!defaultWidths.includes(street.width) && (
              <DropdownMenu.RadioItem
                className="dropdown-menu-radio-item"
                value={street.width.toString()}
                style={{ marginTop: '1em' }}
              >
                <DropdownMenu.ItemIndicator className="dropdown-menu-item-indicator">
                  <Icon name="check" />
                </DropdownMenu.ItemIndicator>
                {prettifyWidth(street.width, units, locale)}
              </DropdownMenu.RadioItem>
            )}
          </DropdownMenu.RadioGroup>

          <DropdownMenu.Item
            className="dropdown-menu-item"
            onSelect={handleWidthEntry}
          >
            <FormattedMessage
              id="width.different"
              defaultMessage="Different widthâ€¦"
            />
          </DropdownMenu.Item>

          <DropdownMenu.Separator className="menu-separator" />

          <DropdownMenu.Label className="dropdown-menu-label">
            <FormattedMessage id="width.units" defaultMessage="Units" />
          </DropdownMenu.Label>
          <DropdownMenu.RadioGroup
            value={units.toString()}
            onValueChange={handleUnitChange}
          >
            <DropdownMenu.RadioItem
              className="dropdown-menu-radio-item"
              value={SETTINGS_UNITS_METRIC.toString()}
            >
              <DropdownMenu.ItemIndicator className="dropdown-menu-item-indicator">
                <Icon name="check" />
              </DropdownMenu.ItemIndicator>
              <FormattedMessage id="width.metric" defaultMessage="Metric" />
            </DropdownMenu.RadioItem>
            <DropdownMenu.RadioItem
              className="dropdown-menu-radio-item"
              value={SETTINGS_UNITS_IMPERIAL.toString()}
            >
              <DropdownMenu.ItemIndicator className="dropdown-menu-item-indicator">
                <Icon name="check" />
              </DropdownMenu.ItemIndicator>
              <FormattedMessage
                id="width.imperial"
                defaultMessage="U.S. customary / imperial"
              />
            </DropdownMenu.RadioItem>
          </DropdownMenu.RadioGroup>

          <DropdownMenu.Arrow className="dropdown-menu-arrow" />
        </DropdownMenu.Content>
      </DropdownMenu.Portal>
    </DropdownMenu.Root>
  )
}

export default StreetMetaWidthNew
