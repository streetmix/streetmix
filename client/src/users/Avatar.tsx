import React from 'react'
import { useGetUserQuery } from '../store/services/api'
import './Avatar.scss'

interface AvatarProps {
  userId: string
}

function Avatar ({ userId }: AvatarProps): React.ReactElement {
  const { data } = useGetUserQuery(userId)

  // Loading and error states will declare an empty `src` property
  return (
    <span className="avatar">
      <img src={data?.profileImageUrl} alt={userId} />
    </span>
  )
}

export default Avatar
