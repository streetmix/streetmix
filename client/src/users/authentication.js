import Cookies from 'js-cookie'
import { jwtDecode } from 'jwt-decode'
import * as Sentry from '@sentry/browser'

import USER_ROLES from '../../../app/data/user_roles.json'
import { app } from '../preinit/app_settings'
import { showError, ERRORS } from '../app/errors'
import { MODES, processMode, getMode, setMode } from '../app/mode'
import { generateFlagOverrides, applyFlagOverrides } from '../app/flag_utils'
import { formatMessage } from '../locales/locale'
import { setPromoteStreet } from '../streets/remix'
import { fetchStreetFromServer, createNewStreetOnServer } from '../streets/xhr'
import store from '../store'
import { updateSettings } from '../store/slices/settings'
import { setSignInData, clearSignInData } from '../store/slices/user'
import { showDialog } from '../store/slices/dialogs'
import { updateStreetIdMetadata } from '../store/slices/street'
import { addToast } from '../store/slices/toasts'
import { getUser, deleteUserLoginToken } from '../util/api'
import { loadSettings } from './settings'

const USER_ID_COOKIE = 'user_id'
const SIGN_IN_TOKEN_COOKIE = 'login_token'
const REFRESH_TOKEN_COOKIE = 'refresh_token'
const LOCAL_STORAGE_SIGN_IN_ID = 'sign-in'

export function doSignIn () {
  store.dispatch(showDialog('SIGN_IN'))
}

export function getSignInData () {
  return store.getState().user.signInData || {}
}

export function isSignedIn () {
  return store.getState().user.signedIn
}

/**
 * Clears sign in data on the client side. Use this when authentication
 * data has become corrupted or expired on the client side and needs to be reset.
 * Do not use this to sign out a user. For that, use signOut(), which ensures
 * that sign out data is also sent to the server.
 */
function clearAllClientSignInData () {
  store.dispatch(clearSignInData())
  window.localStorage.removeItem(LOCAL_STORAGE_SIGN_IN_ID)
  removeSignInCookies()
}

export function onStorageChange () {
  if (isSignedIn() && !window.localStorage[LOCAL_STORAGE_SIGN_IN_ID]) {
    setMode(MODES.FORCE_RELOAD_SIGN_OUT)
    processMode()
  } else if (!isSignedIn() && window.localStorage[LOCAL_STORAGE_SIGN_IN_ID]) {
    setMode(MODES.FORCE_RELOAD_SIGN_IN)
    processMode()
  }
}

function saveSignInDataLocally () {
  const signInData = getSignInData()
  if (signInData) {
    window.localStorage.setItem(
      LOCAL_STORAGE_SIGN_IN_ID,
      JSON.stringify(signInData)
    )
  } else {
    window.localStorage.removeItem(LOCAL_STORAGE_SIGN_IN_ID)
  }
}

function removeSignInCookies () {
  Cookies.remove(SIGN_IN_TOKEN_COOKIE)
  Cookies.remove(REFRESH_TOKEN_COOKIE)
  Cookies.remove(USER_ID_COOKIE)
}

export async function loadSignIn () {
  const signInCookie = Cookies.get(SIGN_IN_TOKEN_COOKIE)
  const refreshCookie = Cookies.get(REFRESH_TOKEN_COOKIE)
  const userIdCookie = Cookies.get(USER_ID_COOKIE)

  if (signInCookie && userIdCookie && refreshCookie) {
    store.dispatch(
      setSignInData({
        token: signInCookie,
        refreshToken: refreshCookie,
        userId: userIdCookie
      })
    )

    saveSignInDataLocally()
  } else if (window.localStorage[LOCAL_STORAGE_SIGN_IN_ID]) {
    // old login data is in localstorage but we don't have the cookies we need
    clearAllClientSignInData()
    setMode(MODES.AUTH_EXPIRED)
    processMode()
    return true
  }

  const signInData = getSignInData()

  // Check if token is valid
  // Note that jwtDecode does not actually validate tokens. We will need
  // another library to do that. For now, we use the try/catch to see if
  // throws an error. If so, we log and report this error so we can see
  // why tokens are invalid, then we force clearing all data so that user
  // can reset and start over.
  try {
    if (signInData.token) {
      jwtDecode(signInData.token)
    }
  } catch (error) {
    Sentry.captureMessage('Error parsing jwt token ', signInData?.token)
    clearAllClientSignInData()
    setMode(MODES.AUTH_EXPIRED)
    processMode()
    return true
  }

  const storage = JSON.parse(window.localStorage.getItem('flags'))
  const sessionOverrides = generateFlagOverrides(storage, 'session')

  let flagOverrides = []

  if (signInData && signInData.token && signInData.userId) {
    const decoded = jwtDecode(signInData.token)

    // Check if token has expired
    const currentDate = new Date()
    const expDate = new Date(decoded.exp * 1000)

    // Set the expiration date to one day early so that tokens
    // don't wait till the last moment to refresh
    expDate.setDate(expDate.getDate() - 1)

    if (currentDate >= expDate) {
      await refreshLoginToken(signInData.refreshToken)
    }
    flagOverrides = await fetchSignInDetails(signInData.userId)
  } else {
    store.dispatch(clearSignInData())
  }

  if (!flagOverrides) {
    flagOverrides = []
  }
  applyFlagOverrides(store.getState().flags, ...flagOverrides, sessionOverrides)

  _signInLoaded()

  return true
}

/**
 *
 * @param {String} refreshToken
 * @returns {Object}
 */
async function refreshLoginToken (refreshToken) {
  const requestBody = JSON.stringify({ token: refreshToken })
  try {
    const response = await window.fetch('/services/auth/refresh-login-token', {
      method: 'post',
      headers: {
        'Content-Type': 'application/json'
      },
      body: requestBody
    })

    if (!response.ok) {
      throw response
    }
  } catch (error) {
    errorRefreshLoginToken(error)
  }
}

function errorRefreshLoginToken (data) {
  if (data.status === 401) {
    signOut(true)

    showError(ERRORS.SIGN_IN_401, true)
    return
  } else if (data.status === 503) {
    showError(ERRORS.SIGN_IN_SERVER_FAILURE, true)
    return
  }

  // Fail silently
  store.dispatch(clearSignInData())
}

/**
 *
 * @param {String} userId
 * @returns {Array}
 */
async function fetchSignInDetails (userId) {
  try {
    // TODO: See if it's possible to use RTK Query's implementation of getUser
    // because that will cache user details.
    const response = await getUser(userId)

    if (response.status !== 200) {
      throw response
    }

    const { flags, roles = [] } = response.data

    const flagOverrides = [
      // all role flag overrides
      ...roles.map((key) =>
        generateFlagOverrides(USER_ROLES[key].flags, `role:${key}`)
      ),
      // user flag overrides
      generateFlagOverrides(flags, 'user')
    ]

    receiveSignInDetails(response.data)
    return flagOverrides
  } catch (error) {
    errorReceiveSignInDetails(error)
  }
}

function receiveSignInDetails (details) {
  const signInData = {
    ...getSignInData(),
    details
  }
  store.dispatch(setSignInData(signInData))
  saveSignInDataLocally()
}

function errorReceiveSignInDetails (data) {
  if (data.status === 401) {
    signOut(true)

    // showError(ERRORS.SIGN_IN_401, true)
    // TODO: Check to make sure that this is the correct place to display
    // this. Currently, this will display for all 401 errors, not just
    // when a valid user who was previously signed in has been signed out.
    store.dispatch(
      addToast({
        message: formatMessage(
          'error.auth-expired',
          'We automatically signed you out due to inactivity. Please sign in again.'
        ),
        component: 'TOAST_SIGN_IN',
        duration: Infinity
      })
    )

    return
  } else if (data.status === 503) {
    showError(ERRORS.SIGN_IN_SERVER_FAILURE, true)
    return
  }

  // Fail silently
  store.dispatch(clearSignInData())
}

export function onSignOutClick (event) {
  signOut(false)

  if (event) {
    event.preventDefault()
  }
}

function signOut (quiet) {
  const signInData = getSignInData()
  store.dispatch(
    updateSettings({
      lastStreetId: null,
      lastStreetNamespacedId: null,
      lastStreetCreatorId: null
    })
  )

  sendSignOutToServer(signInData.userId, quiet)
}

function sendSignOutToServer (userId, quiet) {
  return deleteUserLoginToken(userId)
    .then((response) => {
      if (!quiet) {
        receiveSignOutConfirmationFromServer()
      }
    })
    .catch(errorReceiveSignOutConfirmationFromServer)
    .finally(() => {
      removeSignInCookies()
      window.localStorage.removeItem(LOCAL_STORAGE_SIGN_IN_ID)
    })
}

function receiveSignOutConfirmationFromServer () {
  setMode(MODES.SIGN_OUT)
  processMode()
}

function errorReceiveSignOutConfirmationFromServer () {
  setMode(MODES.SIGN_OUT)
  processMode()
}

function _signInLoaded () {
  loadSettings()
  const street = store.getState().street
  let mode = getMode()

  const surveyStreetId = Cookies.get('last_survey_url')

  // hack to return user to the survey street after signing in
  if (surveyStreetId) {
    Cookies.remove('last_survey_url')
    if (mode === MODES.JUST_SIGNED_IN) {
      window.location = surveyStreetId
    }
  }

  if (
    mode === MODES.CONTINUE ||
    mode === MODES.JUST_SIGNED_IN ||
    mode === MODES.USER_GALLERY ||
    mode === MODES.SURVEY_FINISHED ||
    mode === MODES.GLOBAL_GALLERY
  ) {
    const settings = store.getState().settings

    if (settings.lastStreetId) {
      store.dispatch(
        updateStreetIdMetadata({
          creatorId: settings.lastStreetCreatorId,
          id: settings.lastStreetId,
          namespacedId: settings.lastStreetNamespacedId
        })
      )

      if (mode === MODES.JUST_SIGNED_IN && !street.creatorId) {
        setPromoteStreet(true)
      }

      if (mode === MODES.SURVEY_FINISHED) {
        store.dispatch(
          addToast({
            message: formatMessage(
              'error.survey-finished',
              'Survey complete. Congratulations and thank you!'
            ),
            duration: Infinity
          })
        )
        setMode(MODES.CONTINUE)
      }
      if (mode === MODES.JUST_SIGNED_IN) {
        setMode(MODES.CONTINUE)
      }
    } else {
      setMode(MODES.NEW_STREET)
    }
  }
  mode = getMode()

  switch (mode) {
    case MODES.EXISTING_STREET:
    case MODES.CONTINUE:
    case MODES.USER_GALLERY:
    case MODES.GLOBAL_GALLERY:
      fetchStreetFromServer()
      break
    case MODES.NEW_STREET:
    case MODES.NEW_STREET_COPY_LAST:
      if (app.readOnly) {
        showError(ERRORS.CANNOT_CREATE_NEW_STREET_ON_PHONE, true)
      } else {
        createNewStreetOnServer()
      }
      break
  }
}
