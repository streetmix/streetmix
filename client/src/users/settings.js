import debounce from 'just-debounce-it'
import { MODES, getMode } from '../app/mode'
import store, { observeStore } from '../store'
import { updateSettings } from '../store/slices/settings'
import { setAppFlags } from '../store/slices/app'
import { putUserSettings } from '../util/api'
import { getSignInData, isSignedIn } from './authentication'

const LOCAL_STORAGE_SETTINGS_ID = 'settings'
const SAVE_SETTINGS_DELAY = 500

export function loadSettings () {
  // Get server settings.
  let serverSettings = {}
  const signInData = getSignInData()
  if (isSignedIn() && signInData.details) {
    serverSettings = signInData.details.data
  }

  // Get local settings.
  // Skip this if localStorage is corrupted
  let localSettings = {}
  try {
    if (window.localStorage.getItem(LOCAL_STORAGE_SETTINGS_ID)) {
      localSettings = JSON.parse(
        window.localStorage.getItem(LOCAL_STORAGE_SETTINGS_ID)
      )
    }
  } catch (err) {
    console.error(err)
  }

  // Merge settings to a new object. Server settings take priority and will
  // overwrite local settings.

  const settings = Object.assign({}, localSettings, serverSettings)

  // Except for last street settings -- if we've just signed in, local settings
  // take priority.
  const currentMode = getMode()
  if (currentMode === MODES.JUST_SIGNED_IN) {
    settings.lastStreetId = localSettings.lastStreetId
    settings.lastStreetNamespacedId = localSettings.lastStreetNamespacedId
    settings.lastStreetCreatorId = localSettings.lastStreetCreatorId
  }

  // This is a temporary value used only for the "fetch last street"
  // functionality (this can happen either through the welcome panel)
  // or the /copy-last convenience URL.

  store.dispatch(
    setAppFlags({
      priorLastStreetId: settings.lastStreetId
    })
  )

  // Update our application state.
  store.dispatch(updateSettings(settings))
}

/**
 * Save settings to LocalStorage. This should only be called after the settings
 * reducer has updated. This is meant to mirror application state that
 * should persist across browser sessions and even different users.
 *
 * @param {Object} settings
 */
function saveSettingsLocally (settings) {
  try {
    window.localStorage.setItem(
      LOCAL_STORAGE_SETTINGS_ID,
      JSON.stringify(settings)
    )
  } catch (err) {
    // Ignore localstorage write errors.
  }
}

function saveSettingsToServer (settings) {
  if (!isSignedIn() || store.getState().errors.abortEverything) {
    return
  }

  const userId = getSignInData().userId
  putUserSettings(userId, { data: settings })
}

function persistSettingsToBackends (settings) {
  // Mirror the settings to local storage (so they persist across browser
  // sessions) and also to the server (to a user account, if the user is
  // signed in).
  saveSettingsLocally(settings)
  saveSettingsToServer(settings)
}

/**
 * Use an observer model to set localstorage (a Redux pattern)
 *
 * Similar to:
 * https://egghead.io/lessons/javascript-redux-persisting-the-state-to-the-local-storage
 * https://twitter.com/dan_abramov/status/703684128416333825
 *
 * Benefit: LocalStorage is always reflects the store, no matter how it's
 * updated. Debounced so updates are only made after rapid changes have
 * completed.
 */
export function initSettingsStoreObserver () {
  const select = (state) => state.settings
  const onChange = debounce(persistSettingsToBackends, SAVE_SETTINGS_DELAY)

  return observeStore(select, onChange)
}
