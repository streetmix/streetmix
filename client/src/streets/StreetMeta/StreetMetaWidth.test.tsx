import React from 'react'
import { vi } from 'vitest'
import { within } from '@testing-library/dom'
import { screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'

import { render } from '~/test/helpers/render'
import { updateUnits } from '~/src/users/localization'
import StreetMetaWidth from './StreetMetaWidth'

vi.mock('../../users/localization', () => ({
  updateUnits: vi.fn()
}))

const dummyStreetObject = {
  units: 0,
  width: 10,
  occupiedWidth: 10,
  remainingWidth: 0
}

describe('StreetMetaWidth', () => {
  it('renders', () => {
    const { asFragment } = render(<StreetMetaWidth />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('does not render interactive elements when not editable', async () => {
    const { asFragment } = render(<StreetMetaWidth />, {
      initialState: {
        app: {
          readOnly: true
        }
      }
    })

    expect(screen.queryByRole('button')).toBe(null)
    expect(asFragment()).toMatchSnapshot()
  })

  it('updates units', async () => {
    render(<StreetMetaWidth />)

    // Click the "Change street width" button
    await userEvent.click(screen.getByRole('button'))

    // Wait for dropdown, then change value from dropdown
    await userEvent.click(screen.getByText('U.S. customary / imperial'))

    // The change is made
    expect(updateUnits).toBeCalledWith(1)
  })

  it('renders label with remaining width', () => {
    render(<StreetMetaWidth />, {
      initialState: {
        street: {
          ...dummyStreetObject,
          occupiedWidth: 9,
          remainingWidth: 1
        }
      }
    })

    expect(screen.getByText('1 m room', { exact: false })).toBeInTheDocument()
  })

  it('renders label with over available width', () => {
    render(<StreetMetaWidth />, {
      initialState: {
        street: {
          ...dummyStreetObject,
          occupiedWidth: 11,
          remainingWidth: -1
        }
      }
    })

    expect(screen.getByText('1 m over', { exact: false })).toBeInTheDocument()
  })

  it('renders menu (metric units, default width selected)', async () => {
    render(<StreetMetaWidth />, {
      initialState: {
        street: {
          ...dummyStreetObject,
          width: 18,
          occupiedWidth: 3
        }
      }
    })

    await userEvent.click(screen.getByText('18 m width'))

    expect(
      within(document.querySelector('[aria-checked="true"]')).getByText('18 m')
    ).toBeInTheDocument()
  })

  it('renders menu (metric units, custom width selected)', async () => {
    render(<StreetMetaWidth />, {
      initialState: {
        street: {
          ...dummyStreetObject,
          width: 3,
          occupiedWidth: 3
        }
      }
    })

    await userEvent.click(screen.getByText('3 m width'))

    expect(
      within(document.querySelector('[aria-checked="true"]')).getByText('3 m')
    ).toBeInTheDocument()
  })

  it('renders menu (imperial units, default width selected)', async () => {
    render(<StreetMetaWidth />, {
      initialState: {
        street: {
          ...dummyStreetObject,
          units: 1,
          width: 18.288,
          occupiedWidth: 3.048
        }
      }
    })

    await userEvent.click(screen.getByText('60′ width'))

    expect(
      within(document.querySelector('[aria-checked="true"]')).getByText('60′')
    ).toBeInTheDocument()
  })
})
