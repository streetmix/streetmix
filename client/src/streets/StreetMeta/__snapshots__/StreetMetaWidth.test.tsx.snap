// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StreetMetaWidth > does not render interactive elements when not editable 1`] = `
<DocumentFragment>
  <div
    class="street-meta-item street-meta-item-plain street-width"
  >
    <div
      class="street-meta-icon"
    >
      <svg
        data-icon="ruler"
        data-icon-source="test"
      />
    </div>
    <div
      class="street-meta-content"
    >
      <span
        class="underline"
      >
        0 m width
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`StreetMetaWidth > renders 1`] = `
<DocumentFragment>
  <div
    class="street-meta-item street-width street-width-editable"
  >
    <button
      aria-expanded="false"
      aria-haspopup="menu"
      aria-label="Change width of the street"
      data-state="closed"
      id="radix-:r0:"
      type="button"
    >
      <div
        class="street-meta-icon"
      >
        <svg
          data-icon="ruler"
          data-icon-source="test"
        />
      </div>
      <div
        class="street-meta-content"
      >
        <span
          class="underline"
        >
          0 m width
        </span>
        <svg
          class="menu-carat-down"
          data-icon="chevron-down"
          data-icon-source="test"
        />
      </div>
    </button>
  </div>
</DocumentFragment>
`;
